### ISSUE TITLE
[BUG] Need retry logic around curl in detect_repo

### ISSUE THREAD

**Issue by starbops** [harvester/harvester#3079]
**Describe the bug**

We've spotted that failures might occur around `curl` commands in `detect_repo` during node upgrades due to network issues. This leads to the total failure of the node upgrade Pod because we have `set -e` in the upgrade scripts.

```
node-0:/home/rancher # kubectl -n harvester-system logs hvst-upgrade-52qzw-post-drain-mike-2vl7p
+++ dirname /usr/local/bin/upgrade_node.sh
++ cd /usr/local/bin
++ pwd
+ SCRIPT_DIR=/usr/local/bin
+ source /usr/local/bin/lib.sh
++ UPGRADE_NAMESPACE=harvester-system
++ UPGRADE_REPO_URL=http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso
++ UPGRADE_REPO_VM_NAME=upgrade-repo-hvst-upgrade-52qzw
++ UPGRADE_REPO_RELEASE_FILE=http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
++ UPGRADE_REPO_SQUASHFS_IMAGE=http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/rootfs.squashfs
++ UPGRADE_REPO_BUNDLE_ROOT=http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/bundle
++ UPGRADE_REPO_BUNDLE_METADATA=http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/bundle/metadata.yaml
++ CACHED_BUNDLE_METADATA=
++ HOST_DIR=/host
+ UPGRADE_TMP_DIR=/host/usr/local/upgrade_tmp
+ mkdir -p /host/usr/local/upgrade_tmp
+ case $1 in
+ command_post_drain
+ wait_repo
++ get_repo_vm_status
++ kubectl get virtualmachines.kubevirt.io upgrade-repo-hvst-upgrade-52qzw -n harvester-system '-o=jsonpath={.status.printableStatus}'
+ [[ Running == \R\u\n\n\i\n\g ]]
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
+ echo 'Wait for upgrade repo ready...'
Wait for upgrade repo ready...
+ sleep 5
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
+ echo 'Wait for upgrade repo ready...'
+ sleep 5
Wait for upgrade repo ready...
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
+ echo 'Wait for upgrade repo ready...'
+ sleep 5
Wait for upgrade repo ready...
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
+ echo 'Wait for upgrade repo ready...'
+ sleep 5
Wait for upgrade repo ready...
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
+ echo 'Wait for upgrade repo ready...'
+ sleep 5
Wait for upgrade repo ready...
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
+ echo 'Wait for upgrade repo ready...'
+ sleep 5
Wait for upgrade repo ready...
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
+ echo 'Wait for upgrade repo ready...'
+ sleep 5
Wait for upgrade repo ready...
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
+ echo 'Wait for upgrade repo ready...'
+ sleep 5
Wait for upgrade repo ready...
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
+ echo 'Wait for upgrade repo ready...'
+ sleep 5
Wait for upgrade repo ready...
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
+ echo 'Wait for upgrade repo ready...'
Wait for upgrade repo ready...
+ sleep 5
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml
harvester: v1.1.0
harvesterChart: 1.1.0
os: Harvester v1.1.0
kubernetes: v1.24.7+rke2r1
rancher: v2.6.9
monitoringChart: 100.1.0+up19.0.3
loggingChart: 100.1.3+up3.17.7
kubevirt: 0.54.0-150400.3.3.2
minUpgradableVersion: 'v1.0.3'
rancherDependencies:
  fleet:
    chart: 100.1.0+up0.4.0
    app: 0.4.0
  fleet-crd:
    chart: 100.1.0+up0.4.0
    app: 0.4.0
  rancher-webhook:
    chart: 1.0.6+up0.2.7
    app: 0.2.7
+ detect_repo
++ mktemp --suffix=.yaml
+ release_file=/tmp/tmp.4auA9YNfqw.yaml
+ curl -sfL http://upgrade-repo-hvst-upgrade-52qzw.harvester-system/harvester-iso/harvester-release.yaml -o /tmp/tmp.4auA9YNfqw.yaml
```

**To Reproduce**
1. Do upgrade on the cluster of any number of nodes
2. Do something to break the connection between the upgrade-repo VM and the apply-manifest or upgrade-node Pod before `detect_repo` is being called, e.g. remove the Service resource of the upgrade-repo
3. The upgrade will be marked as failed

**Expected behavior**

Should do retries until the curl commands succeed.

**Support bundle**

**Environment**
 - Harvester ISO version: 
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630):

**Additional context**

**Comment by bk201** [comment]
@starbops I guess we also need this on squash image fetching:
- https://github.com/harvester/harvester/blob/a4872361145e0d0b99c4afeb6052f344acd48f30/package/upgrade/upgrade_node.sh#L426
- https://github.com/harvester/harvester/blob/a4872361145e0d0b99c4afeb6052f344acd48f30/package/upgrade/upgrade_node.sh#L477

**Comment by starbops** [comment]
@bk201 yeah, that's a good catch! There are multiple places that use `curl` to download crucial files in the upgrade scripts. Maybe we should drag them out to form a `download` function or try to leverage some of the retry-related options provided by `curl`. I'll do some tests on that.

**Comment by TachunLin** [comment]
Verified **not fixed** on master-0f590757-head. Move to Implement stage.

### Result
Encounter the `Connection timed out`  after the `curl` retry in the single node post-drain pod
```
Downloading the file from "http://upgrade-repo-hvst-upgrade-5jnvr.harvester-system/harvester-iso/rootfs.squashfs" to "/host/usr/local/upgrade_tmp/tmp.3XEMOhNAMQ"...
+ echo 'Downloading the file from "http://upgrade-repo-hvst-upgrade-5jnvr.harvester-system/harvester-iso/rootfs.squashfs" to "/host/usr/local/upgrade_tmp/tmp.3XEMOhNAMQ"...'
+ curl -sSfL --retry 10 --retry-connrefused http://upgrade-repo-hvst-upgrade-5jnvr.harvester-system/harvester-iso/rootfs.squashfs -o /host/usr/local/upgrade_tmp/tmp.3XEMOhNAMQ
curl: (7) Failed to connect to upgrade-repo-hvst-upgrade-5jnvr.harvester-system port 80: Connection timed out
```

### Test Information
* Test Environment: 1 node1 Harvester local kvm machine
* Harvester version: v1.1.0 with upgrade image contains the chages 

### Verify Steps
1. Prepare a v1.0.3 Harvester cluster with any number of nodes
2. Prepare an upgrade image contains the PR
3. Kickstart the upgrade to v1.1.0 GA 
4. Wait until the upgrade entering phase 4: node upgrade 
5. Check the apply-mainifest job completed `kubectl -n harvester-system get po | grep manifes`
```
harv103upd:~ # kubectl -n harvester-system get po | grep manifest
hvst-upgrade-5jnvr-apply-manifests-d6rzj 
```
6. Copy the output of the upgrade repo VM's Service resource with `kubectl -n harvester-system get svc upgrade-repo-hvst-upgrade-5jnvr -o yaml > upgrade-repo-svc.yaml`
7. Apply the upgrade image contains the PR with `ctr -n k8s.io image import harvester-upgrade-v1.1.0.tar`
8. Check the logs of pos-draind pod `kubectl -n harvester-system get po | grep single` 
Run `kubectl logs hvst-upgrade-5jnvr-single-node-upgrade-harv103upd-wlffk -n harvester-system -f`
9. Resume the upgrade by putting back the Service resource with `kubectl apply -f upgrade-repo-svc.yaml`
10. Confirm upgrade can finish successfully

**Comment by starbops** [comment]
The curl retry is not working in the pod so we're going to fall back to the traditional while loop to do the retry.

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.1.0`. Close this issue.

### Result

* Can trigger th retry logic in 10 seconds interval while unable to download the requested repo file
  ```
  + [[ 9 -lt 100 ]]
  + curl -sSfL http://upgrade-repo-hvst-upgrade-l9stq.harvester-system/harvester-iso/rootfs.squashfs -o /host/usr/local/upgrade_tmp/tmp.P1W3fCiK7o
  curl: (6) Could not resolve host: upgrade-repo-hvst-upgrade-l9stq.harvester-system
  + echo 'Failed to download the requested file from "http://upgrade-repo-hvst-upgrade-l9stq.harvester-system/harvester-iso/rootfs.squashfs" to "/host/usr/local/upgrade_tmp/tmp.P1W3fCiK7o" with error code: 6, retrying (9)...'
  + sleep 10
  Failed to download the requested file from "http://upgrade-repo-hvst-upgrade-l9stq.harvester-system/harvester-iso/rootfs.squashfs" to "/host/usr/local/upgrade_tmp/tmp.P1W3fCiK7o" with error code: 6, retrying (9)...
  + i=10
  + [[ 10 -lt 100 ]]
  + curl -sSfL http://upgrade-repo-hvst-upgrade-l9stq.harvester-system/harvester-iso/rootfs.squashfs -o /host/usr/local/upgrade_tmp/tmp.P1W3fCiK7o
  curl: (6) Could not resolve host: upgrade-repo-hvst-upgrade-l9stq.harvester-system
  Failed to download the requested file from "http://upgrade-repo-hvst-upgrade-l9stq.harvester-system/harvester-iso/rootfs.squashfs" to "/host/usr/local/upgrade_tmp/tmp.P1W3fCiK7o" with error code: 6, retrying (10)...
  + echo 'Failed to download the requested file from "http://upgrade-repo-hvst-upgrade-l9stq.harvester-system/harvester-iso/rootfs.squashfs" to "/host/usr/local/upgrade_tmp/tmp.P1W3fCiK7o" with error code: 6, retrying (10)...'
  ```

* Can resume the download after connection recovered 
  ```
  Failed to download the requested file from "http://upgrade-repo-hvst-upgrade-l9stq.harvester-system/harvester-iso/rootfs.squashfs" to "/host/usr/local/upgrade_tmp/tmp.P1W3fCiK7o" with error code: 6, retrying (18)...
  + i=19
  + [[ 19 -lt 100 ]]
  + curl -sSfL http://upgrade-repo-hvst-upgrade-l9stq.harvester-system/harvester-iso/rootfs.squashfs -o /host/usr/local/upgrade_tmp/tmp.P1W3fCiK7o
  + break
  + shutdown_all_vms
  + kubectl get vmi -A -o json
  + jq -r '.items[] | [.metadata.name, .metadata.namespace] | @tsv'
  + IFS='	'
  + read -r name namespace
  + '[' -z upgrade-repo-hvst-upgrade-l9stq ']'
  + echo 'Stop harvester-system/upgrade-repo-hvst-upgrade-l9stq'
  + virtctl stop upgrade-repo-hvst-upgrade-l9stq -n harvester-system
  Stop harvester-system/upgrade-repo-hvst-upgrade-l9stq
  VM upgrade-repo-hvst-upgrade-l9stq was scheduled to stop
  ```
* And finish the upgrade process
![image](https://user-images.githubusercontent.com/29251855/200242634-3192827b-1180-4929-b64c-c8726ea8f8a9.png)

### Test Information
* Test Environment: 4 nodes Harvester bare machines
* Harvester version: master-0f590757-head

### Verify Steps
Ref to https://github.com/harvester/harvester/issues/3079#issuecomment-1303007551
